### PR TITLE
Tightening validation + adding basic validation tests

### LINKED ISSUE #767: Stricter regex-based validation
**What would you like to be added**:

Add regex based validation in the API: https://github.com/kubernetes-sigs/gateway-api/pull/754#discussion_r687174710

**Why this is needed**:

Decreasing the surface area of the API further. We can relax (not a breaking change) these validations later on when need be.

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This adds regex validation and map keys throughout the v1alpha2 API, consolidates shared types, and adds some very simple validation tests. Much of this logic was pulled from [upstream Kubernetes validation](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go).

**Which issue(s) this PR fixes**:
Fixes #767

**Does this PR introduce a user-facing change?**:
```release-note
Many string fields are now validated with regular expressions.
```

/cc @hbagdi @Miciah @bowei

### CODE REVIEW

File: apis/v1alpha2/shared_types.go [hunk 1]
```diff
@@ -249,4 +238,41 @@ type RouteStatus struct {
...
+type Kind string
+
+// Namespace refers to a Kubernetes namespace. It must be a RFC 1123 label.
+//
+// +kubebuilder:validation:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=63
+type Namespace string
+
+// SectionName is the name of a section in a Kubernetes resource. It must be a
```
Comment (howardjohn) [inline apis/v1alpha2/shared_types.go:264]: I don't think this should be restrictive. Its an extension point, section names in custom types may not have such strict validation

File: apis/v1alpha2/shared_types.go [hunk 2]
```diff
@@ -249,4 +238,41 @@ type RouteStatus struct {
 //
 // +kubebuilder:validation:MinLength=1
 // +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Pattern=`^(\*.)?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
```
Comment (howardjohn) [inline apis/v1alpha2/shared_types.go:241]: This matches IPv4 but not IPv6. Spec declares IPs are not allowd at all though..

File: hack/invalid-examples/httproute/invalid-path.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: invalid-path
+spec:
+  rules:
+  - matches:
+    - path:
+        type: Prefix
+        value: /<>invalid
```
Comment (howardjohn) [inline hack/invalid-examples/httproute/invalid-path.yaml:10]: Why is this invalid? `<>` is valid in a url

File: hack/invalid-examples/gateway/duplicate-listeners.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+kind: Gateway
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: duplicate-listeners
+spec:
+  gatewayClassName: acme-lb
+  listeners:
+  - name: same
```
Comment (howardjohn) [inline hack/invalid-examples/gateway/duplicate-listeners.yaml:8]: how is this validated?

Comment (robscott) [reply hack/invalid-examples/gateway/duplicate-listeners.yaml:8]: The wonders of `+listMapKey=name`

File: apis/v1alpha2/shared_types.go (hunk 1 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:264]: Any particular suggestions for loosening this? This was just pulled from the upstream validation for an RFC 1123 subdomain that's used pretty widely throughout k8s. That seemed like a fairly generic starting point which we can always loosen later if necessary.

File: apis/v1alpha2/shared_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:241]: Is there a regex that could allow all valid hostnames but still invalidate IPv4 addresses? For reference, this is pulled from the upstream kubernetes wildcard subdomain validation (with the exception that the leading `*.` is optional here.

File: hack/invalid-examples/httproute/invalid-path.yaml (hunk 3 above)
Comment (robscott) [reply hack/invalid-examples/httproute/invalid-path.yaml]: Wouldn't this be an invalid path segment per the RFC? Regardless, this was a good reminder that my validation was too strict here simply because we allow RegExp and ImplementationSpecific matching here. I've removed this.

File: apis/v1alpha2/shared_types.go [hunk 5]
```diff
@@ -249,4 +238,41 @@ type RouteStatus struct {
...
 type Hostname string
+
+// Group refers to a Kubernetes Group.
+//
+// +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Pattern=`^$|^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+type Group string
```
Comment (hbagdi) [inline apis/v1alpha2/shared_types.go:248]: I wish this was defined in upstream and we could simply pull it from there. Some other day. No action required.

File: apis/v1alpha2/policy_types.go [hunk 6]
```diff
@@ -23,16 +23,10 @@ package v1alpha2
...
-	// +kubebuilder:validation:MaxLength=253
-	Group string `json:"group"`
+	Group Group `json:"group"`
 
 	// Kind is kind of the target resource.
-	//
-	// +kubebuilder:validation:MinLength=1
-	// +kubebuilder:validation:MaxLength=253
-	Kind string `json:"kind"`
+	Kind Kind `json:"kind"`
```
Comment (hbagdi) [inline apis/v1alpha2/policy_types.go:29]: rant: Why can't github highlight the specific text that has changed in this line.

File: apis/v1alpha2/shared_types.go (hunk 1 above)
Comment (hbagdi) [reply apis/v1alpha2/shared_types.go:264]: Start small and tight, loosen when needed.

File: apis/v1alpha2/httproute_types.go [hunk 7]
```diff
@@ -696,30 +706,34 @@ type HTTPRequestRedirect struct {
 	//
 	// +optional
 	// +kubebuilder:validation:Enum=HTTP;HTTPS
+
```
Comment (hbagdi) [inline apis/v1alpha2/httproute_types.go:709]: we do not usually have this line jump

File: apis/v1alpha2/httproute_types.go [hunk 8]
```diff
@@ -461,14 +466,20 @@ type HTTPRouteMatch struct {
 	// ANDed together, meaning, a request must match all the specified headers
 	// to select the route.
 	//
+	// +listType=map
+	// +listMapKey=name
 	// +optional
+	// +kubebuilder:validation:MaxItems=16
 	Headers []HTTPHeaderMatch `json:"headers,omitempty"`
```
Comment (hbagdi) [inline apis/v1alpha2/httproute_types.go:473]: That is a lot of parameters to base a match on practically. No need to change, I don't feel strongly enough to ask for a change.

**Review by hbagdi — COMMENTED** [review]
I'm really bad with regex so I didn't review them. I'm worse with regex than a person who knows nothing about them.
This PR looks good otherwise.
The deal with such PRs is that to review them you have to look at the part of the codebase that _hasn't_ changed. Will improve iteratively if something is missed.

Thanks for pulling this in last minute.

File: apis/v1alpha2/httproute_types.go (hunk 8 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:473]: Yeah I almost started with something lower, but didn't have a great reason. We use 16 for a lot of similar limits so it seemed like a relatively non-controversial starting point.

File: apis/v1alpha2/httproute_types.go (hunk 7 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Updated, thanks!

File: apis/v1alpha2/httproute_types.go [hunk 9]
```diff
@@ -320,6 +321,13 @@ const (
 	HeaderMatchImplementationSpecific HeaderMatchType = "ImplementationSpecific"
 )
 
+// HTTPHeaderName is the name of an HTTP header.
+//
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=256
+// +kubebuilder:validation:Pattern=`^[-A-Za-z0-9]+$`
```
Comment (howardjohn) [inline apis/v1alpha2/httproute_types.go:328]: this forbids pseudo headers and a bunch of other valid things like periods, slashes, percent signs, etc. rfc7230 3.2.6 describes the full list

Comment (robscott) [reply apis/v1alpha2/httproute_types.go:328]: Interesting, are you able to understand from the RFC which characters are valid for a header name? It's hard for me to distinguish, since that specific section seems to be focused on the field value? If the `token` represents the name, it seems like the allowed set of chars would be a-z0-9 + this:

```
"!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" /  "." / "^" / "_" / "`" / "|" / "~" 
```

Then to further complicate things, the spec seems to explicitly rule out `:` as a value, which seems like it would be an issue for pseudo headers...

Comment (robscott) [reply apis/v1alpha2/httproute_types.go:328]: For reference, here's where the current regex is coming from: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L388-L400. You're completely right that it would prevent pseudo headers though.

Comment (howardjohn) [reply apis/v1alpha2/httproute_types.go:328]: I think all of those are valid.

The trick for pseudo header is that `:` is also the separator for the k/v - so in some cases it ends up being treated as `""` as the key. Not sure if its standard though: 
```
$ curl httpbin.org/get -H ':foo: bar'
{
  "args": {},
  "headers": {
    "": "foo: bar"
```

Comment (howardjohn) [reply apis/v1alpha2/httproute_types.go:328]: Also the RFC is hard to follow but if you look at https://datatracker.ietf.org/doc/html/rfc7230#section-3.2, a header key is a `token`, and a token is one of those you listed above. Looking at the golang source was easier than the RFC...

Comment (robscott) [reply apis/v1alpha2/httproute_types.go:328]: So to clarify, you're saying that we wouldn't need `:` to be a valid header name character?

Comment (howardjohn) [reply apis/v1alpha2/httproute_types.go:328]: If we want to support only _real_ headers - no `:`.

If we want to support pseudo headers though, then we do want it. So that would allow header match on `:path`, `:scheme`, etc.

I think either one is valid, but we should be deliberate. If all pseudoheaders are representable based on other first class fields, then its plausble we do not allow them

File: apis/v1alpha2/shared_types.go (hunk 2 above)
Comment (youngnick) [reply apis/v1alpha2/shared_types.go:241]: I don't think there is, because IPv4 addresses are still technically valid domain names as well, sadly.

File: apis/v1alpha2/shared_types.go (hunk 1 above)
Comment (youngnick) [reply apis/v1alpha2/shared_types.go:264]: +1 to starting with RFC1123 subdomain first, and seeing if we need to expand out from there.

File: apis/v1alpha2/httproute_types.go (hunk 9 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:328]: I think there's been at least some discussion around pseudo header matching being useful in some cases, but since I don't think we've reached any consensus on that, it seems safer to leave it out with the possibility of adding it later if necessary.

Comment (youngnick) [reply apis/v1alpha2/httproute_types.go:328]: Allowing pseudoheader pros:
- allows people to quickly try out some more advanced routing things, before the API supports it.
Cons:
- less incentive to have the API support it, because you can use pseudoheaders.

I think it would be better long-term to have the functionality explicitly defined rather than do the pseudoheader thing, but it's not a strong preference.

**Review by youngnick — COMMENTED** [review]
Overall LGTM, I'm inclined to err on the side of not allowing functionality we might want to take away later (in favour of explicit config).

File: apis/v1alpha2/httproute_types.go (hunk 9 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Extended regex to include characters listed above, avoided pseudo headers for now as support will be much easier to add than remove in the future.

File: apis/v1alpha2/shared_types.go (hunk 2 above)
Comment (bowei) [reply apis/v1alpha2/shared_types.go:241]: I think trying to encode really complicated things into a regex will be problematic. It becomes quite hard to verify that a given regex does exactly what you want by looking at it. 

I would comment that this validation may not remove all cases and controller will need to check.

File: apis/v1alpha2/shared_types.go (hunk 5 above)
Comment (bowei) [reply apis/v1alpha2/shared_types.go:248]: Do we have a reference as to where this regex is using as the basis for tis definition?

File: apis/v1alpha2/shared_types.go (hunk 1 above)
Comment (bowei) [reply apis/v1alpha2/shared_types.go:264]: Which specific section of rfc1123 is this definition from? can you put a href? I'm searching the rfc for the word subdomain and I can't seem to find it easily.

 Isn't this just a RFC1123 domain name (not sub)? 

Also, we should give examples in each of the comments. People are not going to be able to stare at the regex and figure out what the intent is. This applies to all of the really complex validations.

File: apis/v1alpha2/shared_types.go [hunk 10]
```diff
@@ -249,4 +238,41 @@ type RouteStatus struct {
...
+// SectionName is the name of a section in a Kubernetes resource. It must be a
+// RFC 1123 subdomain.
+//
+// +kubebuilder:validation:Pattern=`^$|^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=253
+type SectionName string
+
+// GatewayController is the name of a Gateway API controller. It must be a
+// domain prefixed path.
```
Comment (bowei) [inline apis/v1alpha2/shared_types.go:273]: Add example

**Comment by bowei** [comment]
LGTM -- but please put examples in each of the docstrings. the regexes are pretty complicated to parse to understand what a generally valid string looks like.

File: apis/v1alpha2/shared_types.go (hunk 5 above)
Comment (hbagdi) [reply apis/v1alpha2/shared_types.go:248]: I added a note for this in one of my favorite issues: https://github.com/kubernetes-sigs/gateway-api/issues/629

File: apis/v1alpha2/shared_types.go (hunk 1 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:292]: Added examples for most of these + references to where they originated from. Due to how the regexes are combined together in upstream source it can be difficult to find the true source of the logic, but it's fairly easy to get the regex used by just applying a resource with an invalid value. For example, here's how I figured out the validation for Group:

```
The CustomResourceDefinition "gateways.gateway.networking.k8s.io" is invalid:
* spec.group: Invalid value: "gateway.networ<king.k8s.io": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

File: apis/v1alpha2/shared_types.go (hunk 10 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go:313]: Good call, added examples throughout.

**Comment by robscott** [comment]
Thanks for the great feedback on this @bowei, @hbagdi, @howardjohn, and @youngnick! I think I've addressed everything, PTAL.

File: apis/v1alpha2/httproute_types.go [hunk 11]
```diff
@@ -262,6 +262,7 @@ type HTTPPathMatch struct {
 	//
 	// +optional
 	// +kubebuilder:default="/"
+	// +kubebuilder:validation:MaxLength=1024
```
Comment (bowei) [inline apis/v1alpha2/httproute_types.go:265]: apparently 2000 is the accepted folk limit around the web

File: apis/v1alpha2/httproute_types.go [hunk 12]
```diff
@@ -282,6 +283,23 @@ const (
 	HeaderMatchImplementationSpecific HeaderMatchType = "ImplementationSpecific"
 )
 
+// HTTPHeaderName is the name of an HTTP header.
+//
+// Valid values include:
+//
+// * Authorization
```
Comment (bowei) [inline apis/v1alpha2/httproute_types.go:290]: put "quotes" to make it clear these are strings

File: apis/v1alpha2/httproute_types.go [hunk 13]
```diff
@@ -282,6 +283,23 @@ const (
 	HeaderMatchImplementationSpecific HeaderMatchType = "ImplementationSpecific"
 )
 
+// HTTPHeaderName is the name of an HTTP header.
```
Comment (bowei) [inline apis/v1alpha2/httproute_types.go:286]: Do we need to say it's case-insensitive?

File: apis/v1alpha2/httproute_types.go [hunk 14]
```diff
@@ -282,6 +283,23 @@ const (
...
+//
+// Valid values include:
+//
+// * Authorization
+// * Set-Cookie
+//
+// Invalid values include:
+//
+// * :method
```
Comment (bowei) [inline apis/v1alpha2/httproute_types.go:295]: quotes, say why it's invalid (":", "/" are not a valid character for header names)

File: apis/v1alpha2/httproute_types.go [hunk 15]
```diff
@@ -423,14 +438,20 @@ type HTTPRouteMatch struct {
...
 	Headers []HTTPHeaderMatch `json:"headers,omitempty"`
 
 	// QueryParams specifies HTTP query parameter matchers. Multiple match
 	// values are ANDed together, meaning, a request must match all the
 	// specified query parameters to select the route.
 	//
+	// +listType=map
+	// +listMapKey=name
 	// +optional
+	// +kubebuilder:validation:MaxItems=16
```
Comment (bowei) [inline apis/v1alpha2/httproute_types.go:454]: Is there any harm to increase this to 30-50? I can see hitting 16 limit, probably not 30

File: apis/v1alpha2/httproute_types.go [hunk 16]
```diff
@@ -659,29 +683,32 @@ type HTTPRequestRedirect struct {
...
+	Port *PortNumber `json:"port,omitempty"`
+
 	// StatusCode is the HTTP status code to be used in response.
 	//
 	// Support: Core
 	//
 	// +optional
 	// +kubebuilder:default=302
-	// +kubebuilder:validation=301;302
+	// +kubebuilder:validation:Enum=301;302
```
Comment (bowei) [inline apis/v1alpha2/httproute_types.go:711]: I feel like we should just include all the valid redirects from here:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages

File: apis/v1alpha2/object_reference_types.go [hunk 17]
```diff
@@ -20,15 +20,10 @@ package v1alpha2
 // referrer.
 type LocalObjectReference struct {
 	// Group is the group of the referent.
```
Comment (bowei) [inline apis/v1alpha2/object_reference_types.go:22]: For example: "core", "networking"

File: apis/v1alpha2/object_reference_types.go [hunk 18]
```diff
@@ -20,15 +20,10 @@ package v1alpha2
...
-	// +kubebuilder:validation:MaxLength=253
-	Group string `json:"group"`
+	Group Group `json:"group"`
 
 	// Kind is kind of the referent.
-	//
-	// +kubebuilder:validation:MinLength=1
-	// +kubebuilder:validation:MaxLength=253
-	Kind string `json:"kind"`
+	Kind Kind `json:"kind"`
```
Comment (bowei) [inline apis/v1alpha2/object_reference_types.go:26]: For example: Service, HTTPRoute

File: apis/v1alpha2/shared_types.go [hunk 19]
```diff
@@ -249,4 +238,90 @@ type RouteStatus struct {
...
+
+// Group refers to a Kubernetes Group. It must either be an empty string or a
+// RFC 1123 subdomain.
+//
+// This validation is based off of the corresponding Kubernetes validation:
+// https://git.k8s.io/apimachinery/pkg/util/validation/validation.go#L208
+//
+// Valid values include:
+//
+// * ""
```
Comment (bowei) [inline apis/v1alpha2/shared_types.go:252]: (empty string)

**Comment by bowei** [comment]
A few minor comments 

LGTM, ping me after changes, I will apply real LGTM

File: apis/v1alpha2/httproute_types.go (hunk 11 above)
Comment (hbagdi) [reply apis/v1alpha2/httproute_types.go:265]: opinion:
Can you imagine anyone using a 2000 character path to do matching though?
If anything, I think we are already generous here.

File: apis/v1alpha2/httproute_types.go (hunk 16 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:711]: Agree that we should add this, but didn't want to expand the scope of this PR so added https://github.com/kubernetes-sigs/gateway-api/issues/793 to track this.

File: apis/v1alpha2/httproute_types.go (hunk 13 above)
Comment (hbagdi) [reply apis/v1alpha2/httproute_types.go:286]: We say in the place where this type is used. I'm not sure what is the right place for it. The place of use seems more correct because that gets added to the CRD. I can argue either way though.

File: apis/v1alpha2/httproute_types.go (hunk 14 above)
Comment (hbagdi) [reply apis/v1alpha2/httproute_types.go:295]: We should explicitly call out that psuedo headers are not allowed.

File: apis/v1alpha2/httproute_types.go (hunk 16 above)
Comment (hbagdi) [reply apis/v1alpha2/httproute_types.go:711]: There was a conformance concern for not adding them in the first place.

File: config/crd/v1alpha2/gateway.networking.k8s.io_httproutes.yaml [hunk 20]
```diff
@@ -1219,6 +1289,7 @@ spec:
                             detached from the Gateway. \n Support: Core"
                           maxLength: 253
                           minLength: 1
+                          pattern: ^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
                           type: string
```
Comment (hbagdi) [inline config/crd/v1alpha2/gateway.networking.k8s.io_httproutes.yaml:1293]: What is the largest CRD yet? We should register in that competition if one exists.

File: apis/v1alpha2/shared_types.go (hunk 19 above)
Comment (hbagdi) [reply apis/v1alpha2/shared_types.go:252]: Also mention that it implies core API group. Not a widely known thing AFAICT.

File: apis/v1alpha2/shared_types.go [hunk 21]
```diff
@@ -249,4 +238,90 @@ type RouteStatus struct {
...
+// +kubebuilder:validation:Pattern=`^[a-zA-Z]([-a-zA-Z0-9]*[a-zA-Z0-9])?$`
+type Kind string
+
+// Namespace refers to a Kubernetes namespace. It must be a RFC 1123 label.
+//
+// This validation is based off of the corresponding Kubernetes validation:
+// https://git.k8s.io/apimachinery/pkg/util/validation/validation.go#L187
+//
+// This is used for Namespace name validation here:
+// https://git.k8s.io/apimachinery/pkg/api/validation/generic.go#L63
```
Comment (hbagdi) [inline apis/v1alpha2/shared_types.go:277]: These line references will get out of date. Opt for only the file or even better, target a specific commit like so:
https://github.com/kubernetes/apimachinery/blob/02cfb53916346d085a6c6c7c66f882e3c6b0eca6/pkg/api/validation/generic.go#L63

Open the above link and press 'y'.

File: apis/v1alpha2/httproute_types.go (hunk 15 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:454]: I hadn't really thought of a use case with that many matches, but I can increase to 32.

Comment (robscott) [reply apis/v1alpha2/httproute_types.go:454]: Actually we're using 16 as a limit in a lot of places, including header matching. Unless there's a clear use case I'd be tempted to leave that as is to minimize the changes here.

File: apis/v1alpha2/httproute_types.go (hunk 13 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go:286]: I think it's more appropriate to specify that it's case insensitive on the match field itself (current state). There's nothing on this type definition that can require that matching is case insensitive.

File: apis/v1alpha2/shared_types.go (hunk 21 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Updated, thanks! Thanks for letting me know about that "y" shortcut, super helpful.

File: apis/v1alpha2/shared_types.go (hunk 19 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Updated, thanks!

File: apis/v1alpha2/object_reference_types.go (hunk 18 above)
Comment (robscott) [reply apis/v1alpha2/object_reference_types.go:27]: Updated, thanks!

File: apis/v1alpha2/object_reference_types.go (hunk 17 above)
Comment (robscott) [reply apis/v1alpha2/object_reference_types.go]: Updated, thanks!

File: apis/v1alpha2/httproute_types.go (hunk 14 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Updated, thanks!

File: apis/v1alpha2/httproute_types.go (hunk 12 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Updated, thanks!

**Comment by robscott** [comment]
Thanks for the last round of feedback @bowei and @hbagdi! I think I've responded to everything, PTAL.

File: apis/v1alpha2/httproute_types.go (hunk 11 above)
Comment (bowei) [reply apis/v1alpha2/httproute_types.go:265]: Maybe? There is data that 2000 is effectively the max which is well known. For the others, yes, we have less info so just have to pick a number. I'm not super tied to increase this -- just that we know that more about what the max is.

File: apis/v1alpha2/httproute_types.go (hunk 13 above)
Comment (bowei) [reply apis/v1alpha2/httproute_types.go:286]: I'm ok with anything as long as the generated documentation is clear.

**Comment by hbagdi** [comment]
Great job Rob!
/hold cancel
